pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -380,14 +387,40 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, pod
 
 	sandboxChanged, attempt, sandboxID := m.podSandboxChanged(pod, podStatus)
 	changes := podContainerSpecChanges{
-		CreateSandbox:     sandboxChanged,
-		SandboxID:         sandboxID,
-		Attempt:           attempt,
-		ContainersToStart: make(map[int]string),
-		ContainersToKeep:  make(map[kubecontainer.ContainerID]int),
-		ContainersToKill:  make(map[kubecontainer.ContainerID]containerToKillInfo),
+		CreateSandbox:        sandboxChanged,
+		SandboxID:            sandboxID,
+		Attempt:              attempt,
+		ContainersToStart:    make(map[int]string),
+		ContainersToKeep:     make(map[kubecontainer.ContainerID]int),
+		InitContainersToKeep: make(map[kubecontainer.ContainerID]int),
+		ContainersToKill:     make(map[kubecontainer.ContainerID]containerToKillInfo),
+	}
+
+	// check the status of init containers.
+	initFailed := false
+	// always reset the init containers if the sandbox is changed.
+	if !sandboxChanged {
+	Containers:
+		// keep all successfully completed containers up to and including the first failing container
+		for i, container := range pod.Spec.InitContainers {
+			containerStatus := podStatus.FindContainerStatusByName(container.Name)
+			switch {
+			case containerStatus == nil:
+				continue
+			case containerStatus.State == kubecontainer.ContainerStateRunning:
+				changes.InitContainersToKeep[containerStatus.ID] = i
+			case containerStatus.State == kubecontainer.ContainerStateExited:
+				changes.InitContainersToKeep[containerStatus.ID] = i
+				if containerStatus.ExitCode != 0 {
```

Comment:
I think this little chunk of code might be easier to read as a "containerStatusFailed" helper it it doesn't already exist.

The logic just boils down to: `if containerStatusFailed(container) { initFailed = true; break } else { containersToKeep = i }` if we have the helper I think, and it also saves us from needing a break label.